File: packages/shared/types/workflow.ts
export type WorkflowStatus = "draft" | "active" | "paused";
export type ExecutionStatus = "running" | "completed" | "failed";
export type TriggerType = "event" | "schedule" | "manual";
export type ActionType =
  | "send_email"
  | "add_tag"
  | "remove_tag"
  | "enroll_course"
  | "add_to_list"
  | "remove_from_list"
  | "wait"
  | "webhook"
  | "update_contact"
  | "create_note"
  | "condition";

export interface Workflow {
  id: number;
  tenant_id: number;
  name: string;
  description: string;
  trigger_type: TriggerType;
  trigger_config: Record<string, unknown>;
  status: WorkflowStatus;
  execution_count: number;
  created_at: string;
  updated_at: string;
  actions?: WorkflowAction[];
}

export interface WorkflowAction {
  id: number;
  tenant_id: number;
  workflow_id: number;
  type: ActionType;
  config: Record<string, unknown>;
  sort_order: number;
  delay_seconds: number;
  condition: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
}

export interface WorkflowExecution {
  id: number;
  tenant_id: number;
  workflow_id: number;
  contact_id: number;
  trigger_event: string;
  status: ExecutionStatus;
  current_step: number;
  log: unknown;
  started_at: string;
  completed_at: string | null;
  created_at: string;
  workflow?: Workflow;
  contact?: { id: number; first_name: string; last_name: string; email: string };
}

File: apps/api/internal/models/workflow.go
package models

import (
	"time"

	"gorm.io/datatypes"
	"gorm.io/gorm"
)

// --- Workflow Automation ---

const (
	WorkflowStatusDraft  = "draft"
	WorkflowStatusActive = "active"
	WorkflowStatusPaused = "paused"

	ExecutionRunning   = "running"
	ExecutionCompleted = "completed"
	ExecutionFailed    = "failed"
)

type Workflow struct {
	ID             uint           `gorm:"primarykey" json:"id"`
	TenantID       uint           `gorm:"index;not null;default:1" json:"tenant_id"`
	Name           string         `gorm:"size:255;not null" json:"name"`
	Description    string         `gorm:"type:text" json:"description"`
	TriggerType    string         `gorm:"size:50;not null" json:"trigger_type"` // event, schedule, manual
	TriggerConfig  datatypes.JSON `gorm:"type:jsonb" json:"trigger_config"`    // e.g. {"event": "purchase.completed"}
	Status         string         `gorm:"size:20;default:'draft'" json:"status"`
	ExecutionCount int64          `gorm:"default:0" json:"execution_count"`
	CreatedAt      time.Time      `json:"created_at"`
	UpdatedAt      time.Time      `json:"updated_at"`
	DeletedAt      gorm.DeletedAt `gorm:"index" json:"-"`

	Actions []WorkflowAction `gorm:"foreignKey:WorkflowID" json:"actions,omitempty"`
}

type WorkflowAction struct {
	ID           uint           `gorm:"primarykey" json:"id"`
	TenantID     uint           `gorm:"index;not null;default:1" json:"tenant_id"`
	WorkflowID   uint           `gorm:"index;not null" json:"workflow_id"`
	Type         string         `gorm:"size:50;not null" json:"type"` // send_email, add_tag, remove_tag, enroll_course, wait, webhook, update_contact, create_note, condition
	Config       datatypes.JSON `gorm:"type:jsonb" json:"config"`
	SortOrder    int            `gorm:"default:0" json:"sort_order"`
	DelaySeconds int            `gorm:"default:0" json:"delay_seconds"`
	Condition    datatypes.JSON `gorm:"type:jsonb" json:"condition"` // for if/else branching
	CreatedAt    time.Time      `json:"created_at"`
	UpdatedAt    time.Time      `json:"updated_at"`
}

type WorkflowExecution struct {
	ID           uint           `gorm:"primarykey" json:"id"`
	TenantID     uint           `gorm:"index;not null;default:1" json:"tenant_id"`
	WorkflowID   uint           `gorm:"index;not null" json:"workflow_id"`
	ContactID    uint           `gorm:"index" json:"contact_id"`
	TriggerEvent string         `gorm:"size:100" json:"trigger_event"`
	Status       string         `gorm:"size:20;default:'running'" json:"status"`
	CurrentStep  int            `gorm:"default:0" json:"current_step"`
	Log          datatypes.JSON `gorm:"type:jsonb" json:"log"`
	StartedAt    time.Time      `json:"started_at"`
	CompletedAt  *time.Time     `json:"completed_at"`
	CreatedAt    time.Time      `json:"created_at"`

	Workflow *Workflow `gorm:"foreignKey:WorkflowID" json:"workflow,omitempty"`
	Contact  *Contact  `gorm:"foreignKey:ContactID" json:"contact,omitempty"`
}
